### PR TITLE
[WIP] Allow getting factories from component Host

### PR DESCRIPTION
**Pull request by tigrannajaryan**
Added GetFactory function to Host interface. This allows components
to fetch factories and create components dynamically. The functionality
will be used by receiver creator that creates and destroys other receivers
as it discovers relevant external endpoints from which these receivers
can get data.

-------

This PR shows 2 different approaches.

The first commit in this PR shows the "minimal change" approach to add the functionality we need: https://github.com/tigrannajaryan/opentelemetry-service/pull/3/commits/967e2ff18136f43a30e2bbba57996d22aa015a69
The slight downside is that `Host.GetFactory` does not return the most narrow type possible for the component kind.

The second commit introduces 4 narrower-typed `Host.GetFactory` functions. However this requires more refactoring and introduces slightly weird set of factory types. We have the following composition for factories:

component.Factory -> component.ReceiverFactory -> receiver.BaseFactory -> receiver.Factory

I am not totally convinced that the second commit is an improvement.

### CODE REVIEW

**Comment by jrcamp** [comment]
I don't have a preference either way. I'd be fine with the simpler one. I'll throw out a third alternative that would require reflection, not sure it's any better:

```go
var factory factory.ReceiverFactory
if err := GetFactory(typeStr, &factory); err != nil {
   return err
}
// use receiver factory
```

**Comment by tigrannajaryan** [comment]
> I don't have a preference either way. I'd be fine with the simpler one. I'll throw out a third alternative that would require reflection, not sure it's any better:
> 
> ```go
> var factory factory.ReceiverFactory
> if err := GetFactory(type, &factory.ReceiverFactory); err != nil {
>    return err
> }
> // use receiver factory
> ```

Thanks for suggesting an alternate, I didn't think of this. This eliminates the need to pass the `kind`, but requires returning an err since there is no other way to indicate factory was not found. Also, I think it requires GetFactory to declare its second parameter as `interface{}`, which means we get even less support from the compiler to catch mistakes.

If we see other benefits we can go this way otherwise let's just stick with the first approach.

File: component/component.go [hunk 1]
```diff
@@ -40,4 +51,23 @@ type Host interface {
...
+	//   func (r MyReceiver) Start(host component.Host) error {
+	//     apacheFactory := host.GetFactory(KindReceiver,"apache").(receiver.Factory)
+	//     receiver, err := apacheFactory.CreateMetricsReceiver(...)
+	//     ...
+	//   }
+	// GetFactory can be called by the component anytime after Start() begins and
+	// until Shutdown() is called. Note that the component is responsible for destroying
+	// other components that it creates.
+	GetFactory(kind Kind, componentType string) Factory
+
```
Comment (bogdandrutu) [inline component/component.go]: Do we need this generic method?

**Review by bogdandrutu — APPROVED** [review]

File: component/component.go (hunk 1 above)
Comment (tigrannajaryan) [reply component/component.go]: This is showing 2 different approach. We either need the generic one or the 4 others.

The 4 methods are the approach that you asked for, but I am not sure it is a good idea. It requires more changes and gains are unclear. The composition of interfaces looks confusing to me.

See commit 1 vs commit 1+2 in this PR.

**Comment by owais** [comment]
Second one doesn't seem to be a great improvement to me either. There could also be helpers that can reduce any type checking boilerplate with the first one. I also like @jrcamp's suggestion to read into a typed pointer.

**Pull request closed without merging**